Boost: Add Cornerstone Pages list to WP speculation rules

donnchawp opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Ref: [#42322](https://github.com/Automattic/jetpack/issues/42322)

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a new Speculation_Rules class that adds the cornerstone pages URLs to the speculation rules
* Load them in the Jetpack Boost class, just below Cornerstone Pages, so they're always on.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3ny-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* You'll need WordPress 6.8-beta2 or newer to properly test this as it uses filters from that version of WordPress, and #42374 will have to be merged too.
* Once that's done, look at your website in a logged out browser and check that the page(s) from your Cornerstone Pages list are in the speculation rules, and they are prerendering with moderate eagerness.
* Run `composer phpunit` to run PHP tests.

donnchawp committed:
Add Cornerstone Pages list to WP speculation rules

donnchawp committed:
changelog

donnchawp commented:
I tested it on a site upgraded to WP 6.8 beta 2 and it worked.

haqadn commented on projects/plugins/boost/app/lib/speculation-rules/Speculation_Rules.php:
```diff
@@ -0,0 +1,77 @@
+			'cornerstone-pages-prerender',
+			array(
+				'source'    => 'list',
+				'urls'      => $cornerstone_urls,
```
Having a list means they will be loaded regardless of the links being in the page or not. Instead, we should use `'source' => 'document'` and `href_match`. Which will prerender only for matching links.

donnchawp replied:
Done.

haqadn commented on projects/plugins/boost/app/lib/speculation-rules/Speculation_Rules.php:
```diff
@@ -0,0 +1,77 @@
+	 * @return array Array of cornerstone page URLs
+	 */
+	private function get_cornerstone_urls() {
+		$cornerstone_urls = jetpack_boost_ds_get( 'cornerstone_pages_list' );
```
We should use `Cornerstone_Utils::get_list()` for this.

donnchawp replied:
Done

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/display_speculation_rules

donnchawp committed:
Remove this contract, it's gone

donnchawp committed:
Add tests for the Speculation_Rules class

donnchawp committed:
Hopefully fix this test so it runs in the GitHub actions

donnchawp committed:
Remove the docblock, this class doesn't exist yet

The WP_Speculation_Rules class will be introduced in WP 6.8 but with
this docblock, the PHAN static analysis fails. Instead of adding a
temporary polyfill, I'll try removing this docblock. We can add it back
later.

donnchawp committed:
Add a WP_Speculation_Rules polyfill for tests

donnchawp committed:
Use the polyfill in the tests

haqadn reviewed (changes requested):
I didn't realize I did not submit my review. 🤦🏼 

I added the comments a few days ago and forgot to submit.

donnchawp committed:
Add URLs as regexes without protocol or hostname.

donnchawp committed:
Use get_list() which fixes trailingslashes

donnchawp committed:
Adjust tests so they work with new edits

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/display_speculation_rules

haqadn reviewed (approved):
Looks good to merge! 👍🏼

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/display_speculation_rules
